src/server/services/ActionStorage.ts

```diff
@@ -32,23 +33,34 @@ export default class ActionStorage {
     });
   }
 
-  async isValid(username: string, addr: string): Promise<boolean> {
-    username = sha256(username);
-    addr = sha256(addr);
-
-    const totalUsername = await this.query(username);
-    const totalAddr = await this.query(addr);
-
-    return Number(totalUsername) < LIMIT_USERS && Number(totalAddr) < LIMIT_ADDRESSES;
+  async isValid(opts: { username?: string; addr?: string; ip?: string }): Promise<boolean> {
```

Comment:
it looks a bit ambiguous now to me. if nothing is provided, then it's valid?

here https://github.com/paritytech/substrate-matrix-faucet/pull/188/files#diff-ad701bbca301460f45616bd9f07dd4cca0c72c3ed856701d5e9ed9ebee910100R31 if sender & address wasn't provided, it will be `isValid === true`?